### ISSUE TITLE
Issues with global scope prefixes

### ISSUE THREAD

**Issue by GaryJones** [WordPress-Coding-Standards/WordPress-Coding-Standards#915]
Checked against the Genesis Framework with the `develop` WPCS branch and:

```
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
    <properties>
        <property name="prefixes" type="array" value="genesis" />
    </properties>
</rule>
```
```
phpcs --sniffs=WordPress.NamingConventions.PrefixAllGlobals
```

The following appear to be false positives:

- `Hook names invoked by a theme/plugin should start with the theme/plugin prefix. Found: "'genesis_cpt_archive_settings_capability_' . $this->post_type->name".`.
  Includes a dynamic part of the hook name the reason?
- `Hook names invoked by a theme/plugin should start with the theme/plugin prefix. Found: "$filter_location".` (or `$args['before_sidebar_hook']`).
  A filter name that is completely dynamic.
- `Hook names invoked by a theme/plugin should start with the theme/plugin prefix. Found: "the_search_query".` (or `loginout`, or `widget_title`) .
  Genesis invokes a WordPress filter.
- `Functions declared by a theme/plugin should start with the theme/plugin prefix. Found: "mb_strpos".` (and three other `mb_*()` functions).
  Genesis adds these for backcompat, but they are names that match PHP functions.

**Comment by jrfnl** [comment]
@GaryJones Could you give me a little more context, i.e. code snippets I can use to test with for the first two ?

The last two cases are typical ones for which the whitelist comment was added to the sniff as documented both in the PR as well as in the sniff and the unit test file.

**Comment by jrfnl** [comment]
Just thinking regarding the last one: I could reduce the false positives created by backfills for new PHP functions by checking whether a function of the same name exists.

Caveats:
* This will only work for PHP functions, not for backfills of WP functions.
* The results will depend on the PHP version the sniff is being run on, i.e. if the function was introduced in PHP 7, but the sniffs are run using PHP 5.6, you will still get the false positive.

**Comment by GaryJones** [comment]
> Could you give me a little more context, i.e. code snippets I can use to test with for the first two ?

1.

```php
class Genesis_Admin_CPT_Archive_Settings extends Genesis_Admin_Boxes {
	public function __construct( $post_type ) {
		...
		$menu_ops = array(
			'submenu' => array (
				'parent_slug' => 'edit.php?post_type=' . $this->post_type->name,
				'page_title'  => apply_filters( 'genesis_cpt_archive_settings_page_label', __( 'Archive Settings', 'genesis' ) ),
				'menu_title'  => apply_filters( 'genesis_cpt_archive_settings_menu_label', __( 'Archive Settings', 'genesis' ) ),
				'capability'  => apply_filters( 'genesis_cpt_archive_settings_capability_' . $this->post_type->name, 'manage_categories' ),
			)
		);
		...
	}
	...
}
```

2.
```php
function genesis_get_nav_menu( $args = array() ) {
	...
	$filter_location = 'genesis_' . $sanitized_location . '_nav';

	// Handle back-compat for primary and secondary nav filters.
	if ( 'primary' === $args['theme_location'] ) {
		$filter_location = 'genesis_do_nav';
	} elseif ( 'secondary' === $args['theme_location'] ) {
		$filter_location = 'genesis_do_subnav';
	}

	/**
	 * Filter the navigation markup.
	 *
	 * @since 2.1.0
	 *
	 * @param string $nav_output Opening container markup, nav, closing container markup.
	 * @param string $nav Navigation list (`<ul>`).
	 * @param array $args {
	 *     Arguments for `wp_nav_menu()`.
	 *
	 *     @type string $theme_location Menu location ID.
	 *     @type string $container Container markup.
	 *     @type string $menu_class Class(es) applied to the `<ul>`.
	 *     @type bool $echo 0 to indicate `wp_nav_menu()` should return not echo.
	 * }
	 */
	return apply_filters( $filter_location, $nav_output, $nav, $args );
}
```

They can both be distilled down to simple cases:

```php
$acronym_var = 'acronym';
do_action( "acronym_{$acronym}" ); // OK.
apply_filters( "acronym_{$acronym}" ); // OK.
do_action( 'acronym_' . $acronym ); // OK.
apply_filters( 'acronym_' . $acronym ); // OK.
do_action( $acronym ); // OK.
apply_filters( $acronym ); // OK.

class Acronym_Dynamic_Hooks {
	$filter = 'acronym';
	public function test() {
		do_action( "acronym_{$this->filter}" ); // OK.
		apply_filters( "acronym_{$this->filter}" ); // OK.
		do_action( 'acronym_' . $this->filter ); // OK.
		apply_filters( 'acronym_' . $this->filter ); // OK.
		do_action( $this->filter ); // OK.
		apply_filters( $this->filter ); // OK.
	}
}
```

**Comment by jrfnl** [comment]
@GaryJones Thanks for the examples. Looking into this separately from the backfill whitelisting.

Recognizing this is of course a bit more complicated than just the examples you gave - think:
* prefix in a property
* prefix in a property with an underscore
* prefix in a class constant
* prefix in a class constant with an underscore
* prefix in a class property but assigned from the constructor or elsewhere in the class
etc

(just to name a few of the possible situations...)

**Comment by jrfnl** [comment]
@GaryJones I've been looking into this and would like a second opinion on the direction to take.

> ```php
> do_action( "acronym_{$acronym_filter_var}" ); // OK.
> apply_filters( "acronym_{$acronym_filter_var}" ); // OK.
> do_action( 'acronym_' . $acronym_filter_var ); // OK.
> apply_filters( 'acronym_' . $acronym_filter_var ); // OK.
> ```

:white_check_mark:  In the local branch I'm working on, these are now correctly recognized.

> ```php
> do_action( $acronym_filter_var ); // OK.
> apply_filters( "{$acronym_filter_var}_hook ); // OK.
> ```

:question:  But what to do with completely dynamic hook names or hook names starting with a dynamic part ?
And for that matter, when a dynamic array key is found in `$GLOBALS` ?

```php
$GLOBALS[ $something ] = 'value';
$GLOBALS[ "{$something}_something" ] = 'value';
${$variable} = 'value';
```

At this moment, dynamic array keys in `$GLOBALS` are just ignored and variable variables are not taken into account at all.

The way I read it, your code sample above regarding the dynamic hook names suggests, searching up in the scope for a variable definition, evaluating the value of that and if that starts with (one of) the prefix(es) requested, ignoring it.

So what to do if it doesn't ? Throw a warning ? And in that case, should the same logic be applied to `$GLOBALS` with dynamic keys and variable variables ?

Or should any hook name which is/starts with a dynamic part be ignored by this sniff ?

**Comment by GaryJones** [comment]
> Or should any hook name which is/starts with a dynamic part be ignored by this sniff ?

I'd disagree on that. For instances, Genesis also has instances of:

> Hook names invoked by a theme/plugin should start with the theme/plugin prefix. Found: "{$this->pagehook}_settings_page_form".

Where the pagehook value could come from Genesis or from a child theme (Genesis provides class for creating admin pages). 

Simply ignoring hook names that start with a dynamic part would potentially be missing cases.

I guess the problem is, you can't find out what the value of the variable actually is?

**Comment by jrfnl** [comment]
No, the question is more: how far do we want to go ? The more in-depth processing we do, the slower the sniff will become, while there is also a whitelist comment which can be used if necessary.

It's also a question of consistency - if we try and determine the value of a variable in a hook name, we should do the same with a variable array key for `$GLOBALS` or a variable variable.

And then we would still be left with things we wouldn't be able to resolve, like properties defined in a parent class and used in the child, globals defined in another file etc.

So we basically have a choice to make between these options:
* Blanket ignore both variable names as well as hook names which are completely dynamic or start with a dynamic part.
* Downgrade to `warning` in those cases, but throw the message.
* Try and resolve the variable with all the issues associated with it. If the variable contains/starts with the prefix, do nothing. If it doesn't or can't be resolved, throw a warning anyway.

**Comment by GaryJones** [comment]
> Downgrade to warning in those cases, but throw the message.

For practicality, I think this is the most workable solution right now.
